### PR TITLE
Added scoring

### LINKED ISSUE #31: Système de pointage
* Détruire un bloc (+1)
* Tuer un joueur (+20)
* Mourir (-10)
* Last survivor (+50)

### PR DESCRIPTION
**Pull request by asimmon**
Closes #31

### CODE REVIEW

File: src/frontend/src/game/gameHud.ts [hunk 1]
```diff
@@ -28,11 +27,48 @@ export class GameHud extends GameContainer {
...
+      ratios[playerId] = ratio;
+    }
+
+    const container = new Container();
+    let lastY = 0;
+
+    for (const playerId in this.state.players) {
+      ratios[playerId] = ratios[playerId] - extraRatios / playerCount;
+
+      const something = new Graphics();
```
Comment (joseph-passineau) [inline src/frontend/src/game/gameHud.ts]: Naming thing is hard ;p

File: src/frontend/src/game/gameHud.ts [hunk 2]
```diff
@@ -28,11 +27,48 @@ export class GameHud extends GameContainer {
...
+        extraRatios += ratio;
+      }
+
+      ratios[playerId] = ratio;
+    }
+
+    const container = new Container();
+    let lastY = 0;
+
+    for (const playerId in this.state.players) {
```
Comment (joseph-passineau) [inline src/frontend/src/game/gameHud.ts]: Pourrait etre mis dans une method ColorizePlayer(player: IPlayer)

File: src/server/state.ts [hunk 3]
```diff
@@ -229,6 +250,13 @@ export class GameState extends Schema implements IGameState {
     }
   }
 
+  private planPlayerColors(): void {
+    if (PLAYER_COLORS.length < 4) throw new Error('Not enough player colors in constants');
```
Comment (joseph-passineau) [inline src/server/state.ts]: Is it necessary? We are the one defining the PLAYER_COLORS constant.

File: src/server/state.ts [hunk 4]
```diff
@@ -71,8 +78,19 @@ export class Player extends Schema implements IPlayer {
...
+
+  @type('int32')
+  color: number = 0xffffff;
+
   @type('boolean')
   hasWon: boolean = false;
+
+  public addScore(deltaScore: number): void {
+    this.score += deltaScore;
+    if (this.score < 0) this.score = 0;
```
Comment (joseph-passineau) [inline src/server/state.ts]: Do we really want to put this check? Having a negative score could be a thing. If I look at games like Smash, they give negative scores.

File: src/server/state.ts [hunk 5]
```diff
@@ -518,8 +560,14 @@ export class GameState extends Schema implements IGameState {
     const explosionChain: IBomb[] = []; // FIFO
     const deletedBombIds = new Set<string>();
     const explosionPositions = new Set<string>();
-    const destroyedBlockPositions = new Set<string>();
-    const hitPlayers = new Set<string>();
+
+    const destroyedBlocks = new EquatableSet((o1: DestroyedBlock, o2: DestroyedBlock) => {
```
Comment (joseph-passineau) [inline src/server/state.ts]: Naming is hard ;p

File: src/server/state.ts [hunk 6]
```diff
@@ -518,8 +560,14 @@ export class GameState extends Schema implements IGameState {
...
+      return o1.x === o2.x && o1.y === o2.y;
+    });
+
+    const playerHits = new EquatableSet((o1: PlayerHit, o2: PlayerHit) => {
```
Comment (joseph-passineau) [inline src/server/state.ts]: Naming

File: src/server/state.ts (hunk 3 above)
Comment (asimmon) [reply src/server/state.ts]: That's true

**Pull request merged**